modules/dnstap/dnstap.c

```diff
@@ -0,0 +1,410 @@
+				&m.response_address,
+				&m.has_response_address,
+				&m.response_port,
+				&m.has_response_port);
+		switch (req->qsource.dst_addr->sa_family) {
+			case AF_INET:
+				m.socket_family = DNSTAP__SOCKET_FAMILY__INET;
+				m.has_socket_family = true;
+				break;
+			case AF_INET6:
+				m.socket_family = DNSTAP__SOCKET_FAMILY__INET6;
+				m.has_socket_family = true;
+				break;
+		}
+	}
+
+	if (dnstap_dt->log_resp_pkt) {
+		const knot_pkt_t *rpkt = req->answer;
+		m.response_message.len = rpkt->size;
+		m.response_message.data = (uint8_t *)rpkt->wire;
+		m.has_response_message = true;
+	}
+
+	/* set query time to the timestamp of the first kr_query
+	 * set response time to now
+	 */
+	if (rplan->resolved.len > 0 ) {
+		struct kr_query *first = rplan->resolved.at[0];
+
+		m.query_time_sec = first->timestamp.tv_sec;
+		m.has_query_time_sec = true;
+		m.query_time_nsec = first->timestamp.tv_sec * 1000;
+		m.has_query_time_nsec = true;
+	}
+
+	/* Response time */
+	m.response_time_sec = now.tv_sec;
+	m.has_response_time_sec = true;
+	m.response_time_nsec = now.tv_sec * 1000;
+	m.has_response_time_nsec = true;
+
+	/* Qname */
+	m.query_zone.data = (uint8_t *)knot_pkt_qname(req->answer);
+	m.query_zone.len = req->answer->qname_size;
+	m.has_query_zone = true;
+
+	/* Create a dnstap Message */
+	Dnstap__Dnstap dnstap = DNSTAP__DNSTAP__INIT;
+	dnstap.type = DNSTAP__DNSTAP__TYPE__MESSAGE;
+	dnstap.message = (Dnstap__Message *)&m;
+
+	/* Pack the message */
+	uint8_t *frame = NULL;
+	size_t size = 0;
+	dt_pack(&dnstap, &frame, &size);
+	if (!frame) {
+		return kr_error(ENOMEM);
+	}
+
+	/* Submit a request to send message to fstrm_iothr*/
+	fstrm_res res = fstrm_iothr_submit(dnstap_dt->iothread, dnstap_dt->ioq, frame, size,
+			fstrm_free_wrapper, NULL);
+	if (res != fstrm_res_success) {
+		DEBUG_MSG("Error submitting dnstap message to iothr\n");
+		free(frame);
+		return kr_error(EBUSY);
+	}
+
+	return ctx->state;
+}
+
+KR_EXPORT
+int dnstap_init(struct kr_module *module) {
+	DEBUG_MSG("loading\n");
+	/* allocated memory for internal data */
+	struct dnstap_data *data = malloc(sizeof(*data));
+	if (!data) {
+		return kr_error(ENOMEM);
+	}
+	memset(data, 0, sizeof(*data));
+
+	/* save pointer to internal struct in module for future reference */
+	module->data = data;
+	return kr_ok();
+}
+
+KR_EXPORT
+int dnstap_deinit(struct kr_module *module) {
+	DEBUG_MSG("unloading\n");
+	struct dnstap_data *data = module->data;
+	/* Free allocated memory */
+	if (data) {
+		fstrm_iothr_destroy(&data->iothread);
+		DEBUG_MSG("fstrm iothread destroyed\n");
+		free(data);
+	}
+	return kr_ok();
+}
+
+/* dnstap_unix_writer returns a unix fstream writer
+ * https://gitlab.labs.nic.cz/labs/knot/blob/master/src/knot/modules/dnstap.c#L159
+ */
+static struct fstrm_writer* dnstap_unix_writer(const char *path) {
+	struct fstrm_unix_writer_options *opt = NULL;
+	struct fstrm_writer_options *wopt = NULL;
+	struct fstrm_writer *writer = NULL;
```

Comment:
It looks okay, but can you use cleanup handlers for opt + wopt instead?
See https://github.com/CZ-NIC/knot-resolver/blob/master/contrib/cleanup.h#L25-L28 and https://github.com/CZ-NIC/knot-resolver/blob/30eb8bd5c58e3230439e0e85aa268e9104d7002e/lib/cache.c#L234-L237